src/httpd.rs

```diff
@@ -173,20 +173,108 @@ impl Default for Configuration {
+    InternalServerError,
+    MethodNotImplemented,
+    VersionNotSupported,
+    BadRequest,
+    Unauthorized,
+    Forbidden,
+    NotFound,
+    MethodNotAllowed,
+    ReqTimeout,
+    LengthRequired,
+    UriTooLong,
+    ReqHdrFieldsTooLarge,
+}
+
+impl From<httpd_err_code_t> for HttpdError {
+    #[allow(non_upper_case_globals)]
+    fn from(from: httpd_err_code_t) -> Self {
+        match from {
+            httpd_err_code_t_HTTPD_500_INTERNAL_SERVER_ERROR => HttpdError::InternalServerError,
+            httpd_err_code_t_HTTPD_501_METHOD_NOT_IMPLEMENTED => HttpdError::MethodNotImplemented,
+            httpd_err_code_t_HTTPD_505_VERSION_NOT_SUPPORTED => HttpdError::VersionNotSupported,
+            httpd_err_code_t_HTTPD_400_BAD_REQUEST => HttpdError::BadRequest,
+            httpd_err_code_t_HTTPD_401_UNAUTHORIZED => HttpdError::Unauthorized,
+            httpd_err_code_t_HTTPD_403_FORBIDDEN => HttpdError::Forbidden,
+            httpd_err_code_t_HTTPD_404_NOT_FOUND => HttpdError::NotFound,
+            httpd_err_code_t_HTTPD_405_METHOD_NOT_ALLOWED => HttpdError::MethodNotAllowed,
+            httpd_err_code_t_HTTPD_408_REQ_TIMEOUT => HttpdError::ReqTimeout,
+            httpd_err_code_t_HTTPD_411_LENGTH_REQUIRED => HttpdError::LengthRequired,
+            httpd_err_code_t_HTTPD_414_URI_TOO_LONG => HttpdError::UriTooLong,
+            httpd_err_code_t_HTTPD_431_REQ_HDR_FIELDS_TOO_LARGE => HttpdError::ReqHdrFieldsTooLarge,
+            _ => unimplemented!("all httpd errors should be covered"),
+        }
+    }
+}
+
+impl Into<httpd_err_code_t> for HttpdError {
+    #[allow(non_upper_case_globals)]
+    fn into(self) -> httpd_err_code_t {
+        match self {
+            HttpdError::InternalServerError => httpd_err_code_t_HTTPD_500_INTERNAL_SERVER_ERROR,
+            HttpdError::MethodNotImplemented => httpd_err_code_t_HTTPD_501_METHOD_NOT_IMPLEMENTED,
+            HttpdError::VersionNotSupported => httpd_err_code_t_HTTPD_505_VERSION_NOT_SUPPORTED,
+            HttpdError::BadRequest => httpd_err_code_t_HTTPD_400_BAD_REQUEST,
+            HttpdError::Unauthorized => httpd_err_code_t_HTTPD_401_UNAUTHORIZED,
+            HttpdError::Forbidden => httpd_err_code_t_HTTPD_403_FORBIDDEN,
+            HttpdError::NotFound => httpd_err_code_t_HTTPD_404_NOT_FOUND,
+            HttpdError::MethodNotAllowed => httpd_err_code_t_HTTPD_405_METHOD_NOT_ALLOWED,
+            HttpdError::ReqTimeout => httpd_err_code_t_HTTPD_408_REQ_TIMEOUT,
+            HttpdError::LengthRequired => httpd_err_code_t_HTTPD_411_LENGTH_REQUIRED,
+            HttpdError::UriTooLong => httpd_err_code_t_HTTPD_414_URI_TOO_LONG,
+            HttpdError::ReqHdrFieldsTooLarge => httpd_err_code_t_HTTPD_431_REQ_HDR_FIELDS_TOO_LARGE,
+        }
+    }
+}
+
+pub struct ErrorHandler {
+    error: HttpdError,
+    // handler: Box<dyn Fn(Request, HttpdError) -> Result<Response>>,
+    handler: unsafe extern "C" fn(req: *mut httpd_req_t, error: httpd_err_code_t) -> esp_err_t,
```

Comment:
The error handler is unsafe. Is there a good reason for that?

Reply:
Semi good reason, the error handler doesn't get the `user_ctx` like the normal handler does, so this has to be a unique function that gets called

> When invoked, the validity of `uri`, `method`, `content_len` and `user_ctx` fields of the httpd_req_t parameter is not guaranteed as the HTTP request may be partially received/parsed.

It should be possible to work around it with a global Mutex or something else